docs/allocations.md

```diff
@@ -0,0 +1,280 @@
+
+```
+git clone https://github.com/brendangregg/FlameGraph
+```
+
+Tip: With Docker, you may want to bind mount the `FlameGraph` repository into the container using
+
+```
+docker run -it --rm \
+           --privileged \
+           -v "/path/to/FlameGraphOnYourMachine:/FlameGraph:ro" \
+           -v "$PWD:PWD" -w "$PWD" \
+           swift:latest
+```
+
+or similar.
+
+
+## Tools
+
+In this guide, we will be using the [Linux `perf`](https://perf.wiki.kernel.org/index.php/Main_Page) tool. If you're struggling to get `perf` to work, have a look at our [information regarding `perf`](linux-perf.md). If you're running in a Docker container, don't forget that you'll need a privileged container. And generally, you will need `root` access, so you may need to prefix the commands with `sudo`.
+
+## Getting a `perf` user probe
+
+In this guide, we will be counting the number of allocations. And the very most allocations from a Swift program (on Linux) will be done through the `malloc` function.
+
+To get information about how many times malloc is counted, we will install a `perf` "user probe" on the malloc function. From then on, there will be an event in `perf` that will fire whenever `malloc` is called.
+
+```bash
+# figures out the path to libc
+libc_path=$(readlink -e /lib64/libc.so.6 /lib/x86_64-linux-gnu/libc.so.6)
+
+# installs a probe on `malloc`
+perf probe -x "$libc_path" malloc
+```
+
+The result (hopefully) looks somewhat like this:
+
+```
+Added new event:
+  probe_libc:malloc    (on malloc in /usr/lib/x86_64-linux-gnu/libc-2.31.so)
+
+You can now use it in all perf tools, such as:
+
+	perf record -e probe_libc:malloc -aR sleep 1
+```
+
+What `perf` is telling you here is that it added a new event called `probe_libc:malloc` which will fire every time `malloc` is called.
+
+Let's first of all confirm that our `probe_libc:malloc` probe actually works and run
+
+    perf stat -e probe_libc:malloc -- bash -c 'echo Hello World'
+    
+which should output something like
+
+```
+Hello World
+
+ Performance counter stats for 'bash -c echo Hello World':
+
+              1021      probe_libc:malloc                                           
+
+       0.003840500 seconds time elapsed
+
+       0.000000000 seconds user
+       0.003867000 seconds sys
+```
+
+Which seems to have allocated 1021 times, great. If that probe fired 0 times, something went wrong.
+
+## Making pretty graphs
+
+After we have confirmed that our user probe on `malloc` works in general, let's dial it up a little. The first thing we'll need is a program that we'd like to analyse the allocations of.
+
+For example, we could analyse a program which does 10 subsequent HTTP requests using [AsyncHTTPClient](https://github.com/swift-server/async-http-client). If you're interested in the full source code, please expand below.
```

Comment:
I wonder if it would be more useful use a real(ish) example (like the `NIOSSLContext` allocations in AHC) so that you can break down the various parts like connection setup and different requests. That would be quite nice because it involves getting rid of the "noise" from arg parser etc. WDYT?

Reply:
@glbrntt Well, we need a "driver" program that does real. I kinda went with the simplest program that uses AHC to do 10 subsequent requests and incidentally this _is_ the program with which I found the `NIOSSLContext` allocs in AHC.